kafka-connect/src/main/java/io/tabular/iceberg/connect/data/RecordRouter.java

```diff
@@ -0,0 +1,194 @@
+      String handlerClass = config.getWriteExceptionHandler();
+      FailedRecordFactory factory =
+          (FailedRecordFactory) DeadLetterUtils.loadClass(failedRecordFactoryClass, loader);
+      factory.configure(config.failedRecordHandlerProperties());
+      WriteExceptionHandler handler =
+          (WriteExceptionHandler) DeadLetterUtils.loadClass(handlerClass, loader);
+      handler.initialize(context, config, factory);
+      baseRecordRouter =
+          new RecordRouter.ErrorHandlingRecordRouter(baseRecordRouter, handler, writers, factory);
+    }
+
+    return baseRecordRouter;
+  }
+
+  private static class ConfigRecordRouter extends RecordRouter {
+    private final List<String> tables;
+    private final WriterManager writers;
+
+    ConfigRecordRouter(WriterManager writers, List<String> tables) {
+      this.tables = tables;
+      this.writers = writers;
+    }
+
+    @Override
+    public void write(SinkRecord record) {
+      // route to all tables
+      tables.forEach(
+          tableName -> {
+            writers.write(tableName, record, false);
+          });
+    }
+  }
+
+  private static class StaticRecordRouter extends RecordRouter {
+    private final String routeField;
+    private final WriterManager writers;
+    private final IcebergSinkConfig config;
+
+    StaticRecordRouter(WriterManager writers, IcebergSinkConfig config) {
+      this.routeField = config.tablesRouteField();
+      this.writers = writers;
+      this.config = config;
+    }
+
+    @Override
+    public void write(SinkRecord record) {
+      String routeValue = extractRouteValue(record.value(), routeField);
+      if (routeValue != null) {
+        config
+            .tables()
+            .forEach(
+                tableName ->
+                    config
+                        .tableConfig(tableName)
+                        .routeRegex()
+                        .ifPresent(
+                            regex -> {
+                              boolean matches;
+                              try {
+                                matches = regex.matcher(routeValue).matches();
+                              } catch (Exception error) {
+                                throw new WriteException.RouteRegexException(error);
+                              }
+                              if (matches) {
+                                writers.write(tableName, record, false);
+                              }
+                            }));
+      }
+    }
+  }
+
+  private static class DynamicRecordRouter extends RecordRouter {
+    private final String routeField;
+    private final WriterManager writers;
+
+    DynamicRecordRouter(WriterManager writers, String routeField) {
+      this.routeField = routeField;
+      this.writers = writers;
+    }
+
+    @Override
+    public void write(SinkRecord record) {
+      String routeValue = extractRouteValue(record.value(), routeField);
```

Comment:
@fqtab to talk offline: I think this is a bug/behavior in the original connector.  Need to discuss breaking change with you.  I've copied existing code but I think it's wrong.